---
title: "IBRA User Guide"
author: "Charlotte Soneson"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{IBRA User Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

IBRA is a package to calculate and visualize performance metrics for ranking and binary assignment methods. A typical use cases could be, for example, comparing methods calling differential expression in gene expression experiments, which could be seen as either a ranking problem (estimating the correct effect size and ordering the genes by significance) or a binary assignment problem (determining which genes are differentially expressed and which are not). 

## Creating a data set
The starting point for an evaluation is typically a collection of feature 'scores' (p-values, adjusted p-values or any other ranking score) and the truth (information about the true status, or the true scores for the features). These values are stored in an IBRAData object. Here, we will start from an example IBRAData object provided with the package.

```{r}
library(IBRA)
load(system.file("extdata", "example_data.Rdata", package = "IBRA"))
class(ibradata)
ibradata
```

Since not all methods have adjusted p-values, we will calculate adjusted p-values from nominal p-values.

```{r}
ibradata <- calculate_adjp(ibradata)
```


## Calculating performance scores


```{r}
ibraperf <- calculate_performance(ibradata, binary_truth = "status", 
                                  cont_truth = "status", splv = "none")
slotNames(ibraperf)
```

## Preparing performance object for plotting

```{r}
ibraplot <- prepare_data_for_plot(ibraperf, colorscheme = "Dark2", facetted = TRUE)
```

## Plotting

```{r, fig.width = 7, fig.height = 4, fig.cap = "", warning = FALSE}
plot_tpr(ibraplot)
```

```{r, fig.width = 7, fig.height = 4, fig.cap = "", warning = FALSE}
plot_fdrtprcurve(ibraplot)
```

## Stratification

```{r, fig.width = 7, fig.height = 5, warning = FALSE}
ibraperf <- calculate_performance(ibradata, binary_truth = "status", 
                                  cont_truth = "status", splv = "gene_expression")
ibraplot <- prepare_data_for_plot(ibraperf, colorscheme = "Dark2", facetted = TRUE)
plot_tpr(ibraplot)
```

## Modifying plots

The plotting functions (except plot_overlap) return ggplot objects. Some basic setting (such as point size, size of the text in the panel headers and the axis ranges) can be set in the plot functions. Others can be modified manually. For example, to reduce the size of the axis labels:

```{r, fig.width = 7, fig.height = 5, warning = FALSE}
library(ggplot2)
pp <- plot_tpr(ibraplot, stripsize = 7.5, pointsize = 3)
pp + theme(axis.text.x = element_text(angle = 90, vjust = 0.5,
                                     hjust = 1, size = 10),
          axis.text.y = element_text(size = 10),
          axis.title.x = element_text(size = 10),
          axis.title.y = element_text(size = 10))
```

Or, to modify the facetting and/or label position:

```{r, fig.width = 7, fig.height = 4, warning = FALSE}
pp + theme(axis.text.x = element_text(angle = 90, vjust = 0.5,
                                     hjust = 1, size = 10),
          axis.text.y = element_text(size = 10),
          axis.title.x = element_text(size = 10),
          axis.title.y = element_text(size = 10),
          legend.position = "bottom") +
  facet_wrap(~splitval, nrow = 1)
```

The colors can either be chosen by one of the pre-defined color palettes, or provided by the user. If the wrong number of colors is provided, IBRA will automatically add or remove colors. 

```{r, fig.width = 7, fig.height = 5, warning = FALSE}
ibraplot <- prepare_data_for_plot(ibraperf, 
                                  colorscheme = c("blue", "green", "yellow", "red", "pink"), 
                                  facetted = TRUE)
pp <- plot_tpr(ibraplot, stripsize = 7.5, pointsize = 3)
pp + theme(axis.text.x = element_text(angle = 90, vjust = 0.5,
                                     hjust = 1, size = 10),
          axis.text.y = element_text(size = 10),
          axis.title.x = element_text(size = 10),
          axis.title.y = element_text(size = 10))
```

The plot_overlap function generates venn diagram(s). It takes additional arguments that will be passed to limma::vennDiagram, e.g. to change the size of the text

```{r, fig.width = 10, fig.height = 5, warning = FALSE}
plot_overlap(ibraplot)
plot_overlap(ibraplot, cex = c(1, 0.7, 0.7))
```

